File: android/app/src/main/java/com/smstool/gateway/network/MessageParser.java
package com.smstool.gateway.network;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

/**
 * Parses incoming WebSocket messages from the backend.
 * Provides utility methods for constructing outgoing messages.
 */
public class MessageParser {
    private static final String TAG = "MessageParser";
    private static final Gson gson = new Gson();

    /**
     * Parse the type of a JSON message.
     */
    public static String getMessageType(String jsonText) {
        try {
            JsonObject json = JsonParser.parseString(jsonText).getAsJsonObject();
            return json.has("type") ? json.get("type").getAsString() : null;
        } catch (Exception e) {
            Log.e(TAG, "Failed to parse message type", e);
            return null;
        }
    }

    /**
     * Parse a complete JSON message into a JsonObject.
     */
    public static JsonObject parseMessage(String jsonText) throws Exception {
        try {
            return JsonParser.parseString(jsonText).getAsJsonObject();
        } catch (Exception e) {
            Log.e(TAG, "Failed to parse JSON message", e);
            throw new Exception("Invalid JSON: " + e.getMessage());
        }
    }

    /**
     * Create a device_info message to send on connection.
     */
    public static String createDeviceInfoMessage(String deviceId, String deviceName,
                                                  String androidVersion, String appVersion) {
        JsonObject msg = new JsonObject();
        msg.addProperty("type", "device_info");
        msg.addProperty("message_id", java.util.UUID.randomUUID().toString());
        msg.addProperty("device_id", deviceId);
        msg.addProperty("device_name", deviceName);
        msg.addProperty("android_version", androidVersion);
        msg.addProperty("app_version", appVersion);
        msg.addProperty("connected_at", java.time.Instant.now().toString());
        return msg.toString();
    }

    /**
     * Create a status_update message for a completed job.
     */
    public static String createStatusUpdateMessage(String jobId, String status, int attempt) {
        return createStatusUpdateMessage(jobId, status, attempt, null, null);
    }

    /**
     * Create a status_update message with error details.
     */
    public static String createStatusUpdateMessage(String jobId, String status, int attempt,
                                                   Integer errorCode, String errorMessage) {
        JsonObject msg = new JsonObject();
        msg.addProperty("type", "status_update");
        msg.addProperty("message_id", java.util.UUID.randomUUID().toString());
        msg.addProperty("job_id", jobId);
        msg.addProperty("status", status);
        msg.addProperty("attempt", attempt);
        if (errorCode != null) {
            msg.addProperty("error_code", errorCode);
        } else {
            msg.add("error_code", com.google.gson.JsonNull.INSTANCE);
        }
        if (errorMessage != null) {
            msg.addProperty("error_message", errorMessage);
        } else {
            msg.add("error_message", com.google.gson.JsonNull.INSTANCE);
        }
        msg.addProperty("timestamp", java.time.Instant.now().toString());
        return msg.toString();
    }

    /**
     * Create a pong message in response to a ping.
     */
    public static String createPongMessage(String pingMessageId) {
        JsonObject msg = new JsonObject();
        msg.addProperty("type", "pong");
        msg.addProperty("message_id", java.util.UUID.randomUUID().toString());
        msg.addProperty("ping_message_id", pingMessageId);
        msg.addProperty("timestamp", java.time.Instant.now().toString());
        return msg.toString();
    }

    /**
     * Extract a field from a JSON message.
     */
    public static String getStringField(JsonObject json, String fieldName, String defaultValue) {
        try {
            if (json.has(fieldName) && !json.get(fieldName).isJsonNull()) {
                return json.get(fieldName).getAsString();
            }
        } catch (Exception e) {
            Log.w(TAG, "Failed to extract field: " + fieldName);
        }
        return defaultValue;
    }

    /**
     * Extract an integer field from a JSON message.
     */
    public static Integer getIntField(JsonObject json, String fieldName) {
        try {
            if (json.has(fieldName) && !json.get(fieldName).isJsonNull()) {
                return json.get(fieldName).getAsInt();
            }
        } catch (Exception e) {
            Log.w(TAG, "Failed to extract int field: " + fieldName);
        }
        return null;
    }
}

File: flutter/lib/network/message_parser.dart
import 'package:uuid/uuid.dart';
import 'dart:convert';

abstract class ServerMessage {
  const ServerMessage();
}

class SmsJobMessage extends ServerMessage {
  final String jobId;
  final String to;
  final String body;
  final int maxRetries;

  const SmsJobMessage({
    required this.jobId,
    required this.to,
    required this.body,
    required this.maxRetries,
  });

  factory SmsJobMessage.fromJson(Map<String, dynamic> json) => SmsJobMessage(
        jobId: json['job_id'] ?? json['jobId'] ?? '',
        to: json['to'] ?? '',
        body: json['body'] ?? '',
        maxRetries: json['max_retries'] ?? json['maxRetries'] ?? 3,
      );
}

class PingMessage extends ServerMessage {
  final String messageId;

  const PingMessage(this.messageId);

  factory PingMessage.fromJson(Map<String, dynamic> json) =>
      PingMessage(json['message_id'] ?? '');
}

class ErrorMessage extends ServerMessage {
  final String code;
  final String detail;

  const ErrorMessage({
    required this.code,
    required this.detail,
  });

  factory ErrorMessage.fromJson(Map<String, dynamic> json) => ErrorMessage(
        code: json['code'] ?? 'UNKNOWN',
        detail: json['detail'] ?? '',
      );
}

class MessageParser {
  static ServerMessage parseServerMessage(String jsonStr) {
    try {
      final json = jsonDecode(jsonStr) as Map<String, dynamic>;
      final type = json['type'] as String?;

      switch (type) {
        case 'sms_job':
          return SmsJobMessage.fromJson(json);
        case 'ping':
          return PingMessage.fromJson(json);
        case 'error':
          return ErrorMessage.fromJson(json);
        default:
          throw Exception('Unknown message type: $type');
      }
    } catch (e) {
      throw Exception('Failed to parse server message: $e');
    }
  }

  // Build device_info message for server
  static String buildDeviceInfoMessage({
    required String deviceId,
    required String deviceName,
    required int androidVersion,
    String appVersion = '1.0.0',
  }) {
    return jsonEncode({
      'type': 'device_info',
      'message_id': const Uuid().v4(),
      'device_id': deviceId,
      'device_name': deviceName,
      'android_version': androidVersion,
      'app_version': appVersion,
      'connected_at': DateTime.now().toIso8601String(),
    });
  }

  // Build status_update message for server
  static String buildStatusUpdateMessage({
    required String jobId,
    required String status,
    required int attempt,
    int? errorCode,
    String? errorMessage,
  }) {
    return jsonEncode({
      'type': 'status_update',
      'message_id': const Uuid().v4(),
      'job_id': jobId,
      'status': status,
      'attempt': attempt,
      'error_code': errorCode,
      'error_message': errorMessage,
      'timestamp': DateTime.now().toIso8601String(),
    });
  }

  // Build pong message for server
  static String buildPongMessage(String pingMessageId) {
    return jsonEncode({
      'type': 'pong',
      'message_id': const Uuid().v4(),
      'ping_message_id': pingMessageId,
      'timestamp': DateTime.now().toIso8601String(),
    });
  }
}
